src/libasr/codegen/asr_to_llvm.cpp

```diff
@@ -3315,7 +3318,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 llvm::Value* list = tmp;
                 this->visit_expr_wrapper(asr_target0->m_pos, true);
                 llvm::Value* pos = tmp;
-                target = list_api->read_item(list, pos, true);
+                target = list_api->read_item(list, pos, *module, true, enable_bounds_checking);
```

Comment:
I wouldn't pass it to each function. Instead make `enable_bounds_checking` a static data member of `LLVMList` because it is a global flag for the whole compiler so should be shared across all the `LLVMList` objects.

Reply:
Done, any other changes?

Reply:
Hm I don't think that's a good design: that's like having a global variable. What if we are doing multiple compilation (even in parallel) with multiple options? Then this would fail.

I think we should always be passing all input data as arguments whenever possible. We should not use global or static variables.

Reply:
> Hm I don't think that's a good design: that's like having a global variable.

Well, `lpython --enable_bounds_checking abc.py` means that enable bound checking for lists while compiling `abc.py`. Now `static bool enable_bounds_checking` is not a global variable for the compiler, its a shared variable for all the `LLVMList`'s class objects under `lpython` process which is compiling `abc.py`. So its not the same as having a global variable IMO. The point here is that when we repeatedly pass `enable_bounds_checking` as an argument to all the methods of `LLVMList` class then its fairly easy to make mistakes by passing an wrong value to some methods. This would lead to some selective bound checking on some lists and not some other lists, or bound checking enable for read and not for write (just because we passed false by mistake and provided we don't test this right now).

> What if we are doing multiple compilation (even in parallel) with multiple options? Then this would fail.

By in parallel if you mean `lpython --enable_bounds_checking abc.py` and `lpython xyz.py` launched simultaneously in the background, then it wouldn't fail because the two processes are isolated from each other I think. As I said the static member which is added here is shared across `LLVMList`'s objects and not more than that.

If we don't want to use `static bool LLVMList::enable_bounds_checking` then at least we should properly test this PR and shouldn't merge otherwise it will pile up mistakes and it will be hard to debug in the future once we have the ability to test this. So yes I am not completely against passing `enable_bounds_checking` as an argument but then we should make sure that we are not making any mistakes and doing bounds check selectively for some lists even though user wants it to be enabled for all of their lists and other container data structures.

Reply:
We will be compiling a large project in parallel, in the same process. We parse everything to AST (fully parallel), then based on the file dependencies, we compile everything to ASR. There can in principle be different options for different files. My understanding of the static variable is that it will not work, as in one file we set it to true, for another file we set it to false, bit it overwrites the previous true to false, so the first file will now have false, and things will break. Is my understanding correct? If so, we can't use this approach.

Reply:
Its okay not to use static variables but then we shouldn’t merge this PR without being tested for all the list operations. My main concern is piling up of mistakes if we merge this without testing.

Reply:
Yup, we will wait until https://github.com/lcompilers/lpython/issues/1128 is implemented!

Reply:
What’s preventing it from being implemented? Any blockers? If not then may be let’s go ahead and implementing it.

Reply:
> There can in principle be different options for different files. My understanding of the static variable is that it will not work, as in one file we set it to true, for another file we set it to false, bit it overwrites the previous true to false, so the first file will now have false, and things will break. Is my understanding correct?

If you are launching separate "lpython file1.py", "lpython file2.py" "lpython --enable-bounds-checking file3.py" commands then no, nothing would break because these are separate processes. One process cannot see and overwrite the static variables of a class in another process. So the options you set for one `lpython` command to compile one file won't have any affect on other command with different set of options. In fact, **the change I proposed is limited to the private scope of `LLVMList` and not even classes outside `LLVMList` will be able to touch any of its static variables**. So nothing to worry about. But again I provided two options,

1. `static bool LLVMList::enable_bounds_checking` - This helps us in applying the bounds check by setting this variable one line, this is a safe choice as anyways we launch multiple "lpython" commands to compile multiple files and each command is isolated from the other.
2. Pass it as argument - Well, here you will have to make sure that we test each operation properly and there are very good chances of making a mistake, so very strong testing is needed with this approach.

Its okay to follow 2, I don't have any issues.

Reply:
We want to make sure we do not close a door to single process multiple threads. It seems static variables would close the door, so let's not do that?

Reply:
> It seems static variables would close the door, so let's not do that?

I agree with you. I am opening a new issue for proposing these guidelines and we should include it in README.md so that new developers know that we are open to multithreading based parallel compilation. I feel we will be able to achieve sandboxing and safety in this setting as well.

Reply:
In fact multi-threading based approach will avoid all kinds of dependencies on how fast the build system is because we will parallelise the build on our own. So that's good thing I feel.